# --- Packages --- #

library(DESeq2)

library(ggplot2)

library(EnrichmentBrowser)

# ---------------- #

# ----- DATA ----- #

# Count table generated by featureCounts
exp_data <- read.table(file = "featureCounts.txt",header = T)

# Information table about genes from NCTC8325 strain
gene_inf_data <- read.table(file = "/scripts/GeneSpecificInformation_NCTC8325.tsv", sep = "\t",header = T, na.strings = "", fill = T)

gene_inf_data <- gene_inf_data[-nrow(gene_inf_data),]

# Remove rows with NA values in Gene.ID columns
gene_inf_data <- gene_inf_data[-which(is.na(x = gene_inf_data$Gene.ID)),]

# ---------------- #

# ---- DESeq2 ---- #

# Set experimental conditions
# Condition 1 : Control replicate / Condition 2 : Intracellular persister replicate
cond <- factor(rep(2:1,each = 3))

# Keep expression data
cnts <- exp_data[,7:ncol(exp_data)]

# Set row names as gene id from featureCounts
rownames(cnts) <- exp_data$Geneid

# Order DataFrame samples columns
cnts <- cnts[,order(colnames(cnts))]

# Clean rows names
rownames(cnts) <- vapply(rownames(cnts), function(x) substr(x, start = 6, stop = nchar(x)), character(1))

# Remove genes with no expression in all samples
cnts <- cnts[apply(cnts,1,sum) > 0,]

# Create DESeqDataSet from count table with experimental design
dds <- DESeqDataSetFromMatrix(cnts, DataFrame(cond), ~ cond)

# DESeq analyse
dds <- DESeq(dds)

res <- results(dds, independentFiltering = TRUE, pAdjustMethod = "BH")

# Add BaseMean statistiques columns
cnts["logBaseMean"] <- log2(res$baseMean)

# Add LogFoldChange statistiques columns
cnts["logFC"] <- res$log2FoldChange

# Add adjusted p-value statistiques columns
cnts["padj"] <- res$padj

# Add pvalue statistiques columns
cnts["pvalue"] <- res$pvalue

# Create MA Plot
jpeg(filename = "MA-Plot_of_complete_RNA-seq_dataset.jpg")

plotMA(res, main = "MA-Plot of complete RNA-seq dataset", colNonSig = "black", colSig = "red", lty = 2)

dev.off()

jpeg(filename = "Volcano-Plot_of_complete_RNA-seq_dataset.jpg")

ggplot(cnts, aes(x = logFC, y = -log10(pvalue), color = ifelse(padj < 0.05, "Significant", "Not Significant"))) +
  geom_point(alpha = 0.6, size = 1.5) +
  scale_color_manual(values = c("Not Significant" = "grey", "Significant" = "red")) +
  theme_minimal() +
  labs(
    title = "Volcano Plot of complete RNA-seq dataset",
    x = "log2(Fold Change)",
    y = "-log10(p-value)"
  ) +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold")  # Center and bold title
  ) +
  guides(
    color = guide_legend(title = "Pvalue", 
                         override.aes = list(size = 3))  # Adjust legend appearance
  )

dev.off()

# Variance Stabilizing Transformation
vsd <- vst(dds, blind = TRUE)
pca_data <- plotPCA(vsd, intgroup = "cond", returnData = TRUE)

jpeg(filename = "PCA-Plot_of_complete_RNA-seq_dataset.jpg")

# Plot PCA
ggplot(pca_data, aes(PC1, PC2, color = cond)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(title = "PCA of RNA-Seq Data",
       x = paste0("PC1: ", round(100 * attr(pca_data, "percentVar")[1], 1), "% variance"),
       y = paste0("PC2: ", round(100 * attr(pca_data, "percentVar")[2], 1), "% variance")) +
  theme_minimal() +  # Clean theme
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold")  # Center and bold title
  ) +
  guides(
    color = guide_legend(title = "Experimental conditions",
                         override.aes = list(size = 3))  # Adjust legend appearance
  )

dev.off()

# ---------------- #

# --- Enrichment --- #

# Get sao genome sets from KEGG
gene_set <- get.kegg.genesets(download.kegg.pathways("sao"))

# Genes related to translation
translation_genes <- c(gene_set$sao03010_Ribosome,gene_set$`sao00970_Aminoacyl-tRNA_biosynthesis`)

# Keep counting with meta-informations available
cnts_translation <- cnts[which(rownames(cnts) %in% gene_inf_data$locus.tag),]

# Keep counting which are in genes translation set
cnts_translation <- cnts_translation[which(rownames(cnts_translation) %in% translation_genes),]

# Add genes id columns
cnts_translation["Gene.ID"] <- gene_inf_data$Gene.ID[which(gene_inf_data$locus.tag %in% rownames(cnts_translation))]

# DataFrame used for plotting
plot_data <- cnts_translation[,c("logBaseMean","logFC","padj","Gene.ID")]

# Significance colums by adjusted p-value
plot_data["Significance"] = ifelse(plot_data$padj < 0.05, "Significant", "Not Significant")

# Translation pathway columns
plot_data["Translation"] <- ifelse(plot_data$Gene.ID %in% gene_set$`sao00970_Aminoacyl-tRNA_biosynthesis`,"AA-tRNA_synthetase","Ribosome")

# Color used for significance plotting
p_color <- ifelse(plot_data$Significance == "Significant", "red", "grey")

jpeg(filename = "MA-Plot_of_genes_related_to_translation.jpg")

# Create the plot
ggplot(plot_data, aes(x = logBaseMean, y = logFC)) +
  geom_point(aes(color = p_color), shape = 21, fill = p_color) +  # Base points
  geom_point(data = subset(plot_data, Translation == "AA-tRNA_synthetase"), 
             shape = 21, 
             size = 3, 
             color = "black") +  # Black contour for tRNA-synthetase
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") +  # Horizontal line at 0
  scale_x_continuous(limits = c(0, 20), breaks = seq(0, 20, by = 2)) +
  scale_y_continuous(limits = c(-6, 5), breaks = seq(-6, 5, by = 1)) +
  labs(
    title = "MA-Plot of Genes Related to Translation",  # Adjusted title
    x = "Log2 Base Mean",
    y = "Log2 Fold Change",
    color = "Significance"  # Legend for significance
  ) +
  scale_color_manual(
    values = c("grey" = "grey", "red" = "red"),  # Specify actual colors for significance
    labels = c("grey" = "Non-significant", "red" = "Significant")  # Legend labels
  ) +
  theme_minimal() +  # Clean theme
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold")  # Center and bold title
  ) +
  guides(
    color = guide_legend(title = "Adjusted Pvalue", 
                         override.aes = list(size = 3, fill = c("grey", "red")))  # Adjust legend appearance
  )

dev.off()

# ----------------- #