# --- PACKAGES --- #

library(DESeq2)

library(ggplot2)

library(EnrichmentBrowser)

library(pheatmap)

# ---------------- #

# --- FUNCTION --- #

pathway <- function(genes,pathSet){
  res <- c()
  for (gene in genes) {
    index <- 1
    while (index <= length(pathSet) && !(as.character(gene) %in% pathSet[[index]])) {
      index <- index + 1
    }
    if(index > length(gene_set)){
      res <- c(res,"NA")  
    }
    else {
      res <- c(res,substr(names(pathSet)[index],10,nchar(names(pathSet)[index])))
    }
  }
  return(res)
}

# ---------------- #

# ----- DATA ----- #

# Count table generated by featureCounts
exp_data <- read.table(file = "featureCounts.txt",header = T)

# Information table about genes from NCTC8325 strain
gene_inf_data <- read.table(file = "/scripts/GeneSpecificInformation_NCTC8325.tsv", sep = "\t",header = T, na.strings = "", fill = T)

gene_inf_data <- gene_inf_data[-nrow(gene_inf_data),]

# Remove rows with NA values in Gene.ID columns
gene_inf_data <- gene_inf_data[-which(is.na(x = gene_inf_data$Gene.ID)),]

# Get sao genome sets from KEGG
gene_set <- get.kegg.genesets(download.kegg.pathways("sao"))

# Get gene translation factors
gene_translation_factors <- read.table(file = "/scripts/GeneSpecificInformation_SAO_TranslationFactors.tsv",header = T,sep = "\t")

# ---------------- #

# ---- DIFFERENTIAL ANALYSIS ---- #

# Set experimental conditions
# Condition 1 : Control replicate / Condition 2 : Intracellular persister replicate
cond <- factor(rep(2:1,each = 3))

# Keep expression data
cnts <- exp_data[,7:ncol(exp_data)]

# Set row names as gene id from featureCounts
rownames(cnts) <- exp_data$Geneid

# Order DataFrame samples columns
cnts <- cnts[,order(colnames(cnts))]

# Clean rows names
rownames(cnts) <- vapply(rownames(cnts), function(x) substr(x, start = 6, stop = nchar(x)), character(1))

# Remove genes with no expression in all samples
cnts <- cnts[apply(cnts,1,sum) > 0,]

# Create DESeqDataSet from count table with experimental design
dds <- DESeqDataSetFromMatrix(cnts, DataFrame(cond), ~ cond)

# DESeq analyse
dds <- DESeq(dds)

# Get Results from DESeq2
res <- results(dds, independentFiltering = TRUE, pAdjustMethod = "BH")

# Update colnames about samples
names(cnts) <- paste0(sapply((strsplit(colnames(cnts),"_")),"[[",1),"_DEFAULT")

# Bind normalized counts to the DataFrame
cnts <- cbind(cnts,as.data.frame(counts(dds, normalized = TRUE)))

# Update colnames about samples
names(cnts)[7:ncol(cnts)] <- paste0(sapply((strsplit(colnames(cnts)[7:ncol(cnts)],"_")),"[[",1),"_NORM")

# Add BaseMean statistiques columns
cnts["logBaseMean"] <- log2(res$baseMean)

# Add LogFoldChange statistiques columns
cnts["logFC"] <- res$log2FoldChange

# Add adjusted p-value statistiques columns
cnts["padj"] <- res$padj

# Add pvalue statistiques columns
cnts["pvalue"] <- res$pvalue

# Add significance column
cnts["significance"] <- ifelse(cnts$padj < 0.05, "Significant", "Not Significant")

# Add pathway column
cnts["pathway"] <- pathway(rownames(cnts),gene_set)

# Add translation factors
cnts[as.character(gene_translation_factors$GeneID),"pathway"] <- "Translation factors"

# Create MA Plot
jpeg(filename = "MA-Plot_of_complete_RNA-seq_dataset.jpg",width = 1000, height = 1000)

plotMA(res, main = "MA-Plot of complete RNA-seq dataset", colNonSig = "black", colSig = "red", lty = 2)

# Add a legend
legend("bottomleft", legend = c("Significant", "Non-significant"), 
       col = c("red", "black"), pch = c(16, 16), bty = "n")

dev.off()

jpeg(filename = "Volcano-Plot_of_complete_RNA-seq_dataset.jpg",width = 1000, height = 1000)

# Create Volcano Plot
ggplot(cnts, aes(x = logFC, y = -log10(pvalue), color = significance)) +
  geom_point(alpha = 0.6, size = 1.5) +
  scale_color_manual(values = c("Not Significant" = "grey", "Significant" = "red")) +
  theme_minimal() +
  labs(
    title = "Volcano Plot of complete RNA-seq dataset",
    x = "log2(Fold Change)",
    y = "-log10(p-value)"
  ) +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold")  # Center and bold title
  ) +
  guides(
    color = guide_legend(title = "Pvalue", 
                         override.aes = list(size = 3))  # Adjust legend appearance
  )

dev.off()

# Variance Stabilizing Transformation
vsd <- vst(dds, blind = TRUE)
pca_data <- plotPCA(vsd, intgroup = "cond", returnData = TRUE)

jpeg(filename = "PCA-Plot_of_complete_RNA-seq_dataset.jpg",width = 1000, height = 1000)

# Plot PCA by samples conditions
ggplot(pca_data, aes(PC1, PC2, color = cond)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(title = "PCA of RNA-Seq Data",
       x = paste0("PC1: ", round(100 * attr(pca_data, "percentVar")[1], 1), "% variance"),
       y = paste0("PC2: ", round(100 * attr(pca_data, "percentVar")[2], 1), "% variance")) +
  theme_minimal() +  # Clean theme
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),  # Center and bold title
    legend.title = element_text(size=15, face = "bold"),
    legend.text = element_text(size=10),
    legend.key.size = unit(2, 'cm')
  ) +
  guides(
    color = guide_legend(title = "Experimental conditions",
                         override.aes = list(size = 3))  # Adjust legend appearance
  ) +
  scale_color_discrete(name = "Conditions", labels = c("Control","Intracellular persister"))

dev.off()

# PCA by genes

pca_res <- prcomp(na.exclude(cnts[cnts$significance == "Significant" & cnts$pathway != "NA",7:12]), scale. = TRUE)
pca_array <- cbind(data.frame(pca_res$x),pathway = cnts[cnts$significance == "Significant" & cnts$pathway != "NA",]$pathway)

jpeg(filename = "PCA-Plot_of_DEG_RNA-seq_dataset.jpg",width = 1200, height = 1000)

ggplot(pca_array, aes(x = as.numeric(PC1), y = as.numeric(PC2), color = pathway)) +
  geom_point(size = 3, alpha = 0.8) + # Plot genes
  labs(
    title = "PCA of Gene Expression",
    x = paste0("PC1 (", round(100 * summary(pca_res)$importance[2, 1], 1), "% variance)"),
    y = paste0("PC2 (", round(100 * summary(pca_res)$importance[2, 2], 1), "% variance)")
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    legend.position = "right",
    legend.title = element_text(size = 15, face = "bold"),  # Reduce legend title size
    legend.text = element_text(size = 10),   # Reduce legend text size
    legend.key.size = unit(0.5, "cm")    
  )

dev.off()

# HEATMAP

jpeg(filename = "Heatmap_of_DEG_RNA-seq_dataset.jpg",width = 1200, height = 1400)

pheatmap(log2(na.exclude(cnts[cnts$significance == "Significant",7:12]) + 1),
         scale = "row",                     # Normalize rows (genes) for better visualization
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         annotation_col = data.frame(conditions = ifelse(cond == 1, "Control","Intracellular persister"),row.names = colnames(cnts[,7:12])), # Add sample condition annotations
         annotation_row = data.frame(pathway = cnts[cnts$significance == "Significant",]$pathway,row.names = rownames(cnts[cnts$significance == "Significant",])), # Add pathway annotations if available
         show_rownames = FALSE,              # Show gene names
         show_colnames = TRUE,              # Show sample names
         fontsize_row = 5,                  # Adjust row font size if needed
         fontsize_col = 8,                  # Adjust column font size
         color = colorRampPalette(c("blue", "white", "red"))(50),
         main = "DEG Heatmap")  # Color gradient

dev.off()

# ------------------------------- #

# --- Translation Pathway --- #

# Genes related to translation
cnts_translation <- cnts[cnts$pathway == "Ribosome" | cnts$pathway == "Aminoacyl-tRNA_biosynthesis" | cnts$pathway == "Translation factors",]

# Add genes id columns
cnts_translation["Gene.ID"] <- gene_inf_data$Gene.ID[which(gene_inf_data$locus.tag %in% rownames(cnts_translation))]

jpeg(filename = "MA-Plot_of_genes_related_to_translation.jpg",width = 1000, height = 1000)

ggplot(cnts_translation, aes(x = logBaseMean, y = logFC)) +
  geom_point(aes(color = ifelse(cnts_translation$pathway == "Aminoacyl-tRNA_biosynthesis","black","white"), fill = ifelse(cnts_translation$significance == "Significant","red","grey")), shape = 21, size = 3) +  # Base points
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") +  # Horizontal line at 0
  scale_x_continuous(limits = c(0, 20), breaks = seq(0, 20, by = 2)) +
  scale_y_continuous(limits = c(-6, 5), breaks = seq(-6, 5, by = 1)) +
  labs(
    title = "MA-Plot of genes related to translation",  # Adjusted title
    x = "Log2 Base Mean",
    y = "Log2 Fold Change",
    color = "Pathway"  # Legend for significance
  ) +
  scale_color_manual(
    values = c("white" = "white", "black" = "black"),  # Specify actual colors for significance
    labels = c("white" = "Ribosome / Translation Factors", "black" = "Aminoacyl-tRNA_biosynthesis")  # Legend labels
  ) +
  scale_fill_manual(
    values = c("grey" = "grey", "red" = "red"),
    labels = c("grey" = "Non-significant", "red" = "Significant")) +
  theme_minimal() +  # Clean theme
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),  # Center and bold title
    legend.title = element_text(size=15, face = "bold"),
    legend.text = element_text(size=10),
    legend.key.size = unit(2, 'cm')
  ) +
  guides(
    fill = guide_legend(title = "Adjusted Pvalue", 
                         override.aes = list(size = 3, fill = c("grey", "red"), color = "white")),  # Adjust legend appearance
    color = guide_legend(title = "Translation",
                         override.aes = list(size = 3, color = c("black", "white")))
  )

dev.off()

# PCA Plot

# Variance Stabilizing Transformation
vsd_translation <- vsd[rownames(vsd) %in% rownames(cnts_translation), ]

# Recalculate PCA
pca_data_translation <- plotPCA(vsd_translation, intgroup = "cond", returnData = TRUE)

jpeg(filename = "PCA-Plot_of_Translation_RNA-seq_dataset.jpg",width = 1000, height = 1000)

# Plot
ggplot(pca_data_translation, aes(PC1, PC2, color = cond)) +
  geom_point(size = 4) +
  labs(title = "PCA for Translation Pathway",
       x = paste0("PC1: ", round(100 * attr(pca_data_translation, "percentVar")[1], 1), "% variance"),
       y = paste0("PC2: ", round(100 * attr(pca_data_translation, "percentVar")[2], 1), "% variance")) +
  scale_color_discrete(name = "Conditions", labels = c("Control","Intracellular persister")) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),  # Center and bold title
    legend.title = element_text(size=15, face = "bold"),
    legend.text = element_text(size=10),
    legend.key.size = unit(2, 'cm')
  )

dev.off()

# PCA for DEG in translation pathway

pca_res_trans <- prcomp(na.exclude(cnts_translation[cnts_translation$significance == "Significant",7:12]), scale. = TRUE)
pca_array_trans <- cbind(data.frame(pca_res_trans$x),pathway = cnts_translation[cnts_translation$significance == "Significant",]$pathway)

jpeg(filename = "PCA-Plot_of_DEG_Translation_RNA-seq_dataset.jpg",width = 1000, height = 1000)

ggplot(pca_array_trans, aes(x = as.numeric(PC1), y = as.numeric(PC2), color = pathway)) +
  geom_point(size = 3, alpha = 0.8) + # Plot genes
  labs(
    title = "PCA of Gene Expression",
    x = paste0("PC1 (", round(100 * summary(pca_res_trans)$importance[2, 1], 1), "% variance)"),
    y = paste0("PC2 (", round(100 * summary(pca_res_trans)$importance[2, 2], 1), "% variance)")
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    legend.position = "right",
    legend.title = element_text(size = 15, face = "bold"),  # Reduce legend title size
    legend.text = element_text(size = 10),   # Reduce legend text size
    legend.key.size = unit(0.5, "cm")    
  )

dev.off()

jpeg(filename = "PCA-Plot_of_DEG_Translation_RNA-seq_dataset_Facet.jpg",width = 1000, height = 1000)

ggplot(pca_array_trans, aes(x = as.numeric(PC1), y = as.numeric(PC2), color = pathway)) +
  geom_point(size = 3, alpha = 0.8) + # Plot genes
  facet_wrap(~ pathway, scales = "free", ncol = 3) +
  labs(
    title = "PCA of Gene Expression",
    x = paste0("PC1 (", round(100 * summary(pca_re_trans)$importance[2, 1], 1), "% variance)"),
    y = paste0("PC2 (", round(100 * summary(pca_res_trans)$importance[2, 2], 1), "% variance)")
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    legend.position = "right",
    legend.title = element_text(size = 15, face = "bold"),  # Reduce legend title size
    legend.text = element_text(size = 10),   # Reduce legend text size
    legend.key.size = unit(0.5, "cm")    
  )

dev.off()

# Heatmap for DEG in translation pathway

jpeg(filename = "Heatmap_of_DEG_Translation_RNA-seq_dataset.jpg",width = 1200, height = 1400)

pheatmap(log2(cnts_translation[cnts_translation$significance == "Significant",7:12] + 1),
         scale = "row",                     # Normalize rows (genes) for better visualization
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         annotation_col = data.frame(conditions = ifelse(cond == 1, "Control","Intracellular persister"),row.names = colnames(cnts_translation[,7:12])), # Add sample condition annotations
         annotation_row = data.frame(pathway = cnts_translation[(cnts_translation$significance == "Significant"),]$pathway,row.names = rownames(cnts_translation[(cnts_translation$significance == "Significant"),])), # Add pathway annotations if available
         show_rownames = FALSE,              # Show gene names
         show_colnames = TRUE,              # Show sample names
         fontsize_row = 6,                  # Adjust row font size if needed
         fontsize_col = 8,                  # Adjust column font size
         color = colorRampPalette(c("blue", "white", "red"))(50),
         main = "DEG Translation Heatmap")  # Color gradient

dev.off()

# ----------------- #